pkg/controllers/state/cluster.go

```diff
@@ -307,15 +314,81 @@ func (c *Cluster) UpdatePod(ctx context.Context, pod *corev1.Pod) error {
 	return err
 }
 
+// AckPods marks the pod as acknowledged for scheduling from the provisioner. This is only done once per-pod.
+func (c *Cluster) AckPods(pods ...*corev1.Pod) {
+	now := c.clock.Now()
+	for _, pod := range pods {
+		// store the value as now only if it doesn't exist.
+		c.podAcks.LoadOrStore(types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, now)
+	}
+}
+
+// PodAckTime will return the time the pod was first seen in our cache.
+func (c *Cluster) PodAckTime(podKey types.NamespacedName) time.Time {
+	if ackTime, ok := c.podAcks.Load(podKey); ok {
+		return ackTime.(time.Time)
+	}
+	return time.Time{}
+}
+
+// MarkPodSchedulingDecisions keeps track of when we first tried to schedule a pod to a node.
+// This also marks when the pod is first seen as schedulable for pod metrics.
+// We'll only emit a metric for a pod if we haven't done it before.
+func (c *Cluster) MarkPodSchedulingDecisions(podErrors map[*corev1.Pod]error, pods ...*corev1.Pod) {
+	now := c.clock.Now()
+	for _, p := range pods {
+		nn := client.ObjectKeyFromObject(p)
+		// If there's no error for the pod, then we mark it as schedulable
+		if err, ok := podErrors[p]; !ok || err == nil {
+			c.podsSchedulableTimes.LoadOrStore(nn, now)
+		}
+		_, alreadyExists := c.podsSchedulingAttempted.LoadOrStore(nn, now)
+		// If we already attempted this, we don't need to emit another metric.
+		if !alreadyExists {
+			// We should have ACK'd the pod.
+			if ackTime := c.PodAckTime(nn); !ackTime.IsZero() {
+				PodSchedulingDecisionSeconds.Observe(c.clock.Since(ackTime).Seconds(), nil)
+			}
+		}
+	}
+}
+
+// PodSchedulingDecision returns when Karpenter first decided if a pod could schedule a pod in scheduling simulations.
+// This returns 0, false if Karpenter never made a decision on the pod.
+func (c *Cluster) PodSchedulingDecision(podKey types.NamespacedName) time.Time {
```

Comment:
```suggestion
func (c *Cluster) PodSchedulingDecisionTime(podKey types.NamespacedName) time.Time {
```

nit: since this is returning a time -- I thought it was a boolean when I first read the function name